rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,400 @@
+we should benchmark our design against actual UI examples. I think that the
+one used in this RFC as driving example (the RPG tabbed menu) illustrates well
+enough how we could implement Benchmark 2 (main menu) using our system.
+Benchmark 1 is irrelevant. Benchmark 3 could potentially be solved by just
+marking the interactive element with the `Focusable` component and nothing
+else, so that every element can be reached from anywhere.
+
+
+## Prior art
+
+I've only had a cursory glance at what already exists in the domain.
+
+Beside [godot](https://github.com/godotengine/godot-docs/blob/master/tutorials/ui/gui_navigation.rst), I found the [Microsoft
+`FocusManager` documentation](https://docs.microsoft.com/en-us/windows/apps/design/input/focus-navigation-programmatic)
+to be very close to what I am currently doing ([additional Microsoft
+resources](https://docs.microsoft.com/en-us/windows/apps/design/input/focus-navigation))
+
+Our design differs from Microsoft's UI navigation in a few key ways.
+We have game-specific assumptions. For example: we assume a "trail" of active elements
+that leads to the sub-sub-submenu we are currently focused in, we rely on that
+trail to navigate containing menus from the focused element in the submenu.
+
+Otherwise, given my little experience in game programming, I am probably overlooking some
+standard practices. My go-to source ("Game Engine Architecture" by Jason
+Gregory) doesn't mention ui navigation systems.
+
+## Open questions
+
+### Naming
+
+* `NavFence`: If a `NavFence` is pretty much just an abstraction to describe a
+  menu, why not call it a `NavMenu` instead?
+* `sequence menu`: doesn't express it's a menu that can be changed with special
+  `NavRequest::MenuMove` even when focusing in another menu.
+* `NavRequest::MenuMove`: bad naming, given everything is moving within a menu
+* `NavEvent::Caught`: doesn't really express that it's just a `NavRequest` that
+  led to no change in focus
+
+### `NavRequest` system
+
+I wanted to defer the raw input processing to the user. The current
+implementation **requires** the user to specify the input button mapping to the
+`NavRequest`s. However, if we want to provide an out-of-the-box working
+experience, we should have at least a default mapping that can be modified at
+runtime.
+
+I think it makes sense to have a default mapping, as technically it's already
+the case with the `Interaction` component. And having something working
+out of the box is a huge positive.
+
+I am against an inflexible system where it is impossible to change the input
+mappings, any respectable game let you change the input mapping. And if not,
+your players start very quickly to ask for one! So it is necessary to be able
+to remap the buttons.
+
+How would that work? Is there already examples in bevy where we provide default
+button mappings that can be changed?
+
+I think we should think deeply about input handling in default plugins.
```

Comment:
Agreed; this should be subsumed under whatever input mapping standard we agree upon.